Python/Source/Lift.cpp

```diff
@@ -0,0 +1,90 @@
+#include <vector>
+#include "pybind11/functional.h"
+#include "Autobox.hpp"
+#include "Rover/Box.hpp"
+#include "Rover/Evaluator.hpp"
+
+using namespace pybind11;
+
+namespace {
+  using namespace Rover;
+  using ArgsCont = std::vector<Box<object>>;
+  using KwargsCont = std::vector<std::pair<handle,
+    Box<object>>>;
+  
+  ArgsCont args_apply(const std::function<Box<object>(object&&)>& func,
+      tuple&& args) {
+    auto result = ArgsCont();
+    for(auto i = 0U; i < args.size(); ++i) {
+      result.push_back(func(std::move(args[i])));
+    }
+    return result;
+  }
+
+  KwargsCont args_apply(const std::function<Box<object>(object&&)>& func,
+      dict&& args) {
+    auto result = KwargsCont();
+    for(auto& arg : args) {
+      auto value = func(reinterpret_borrow<object>(std::move(arg.second)));
+      result.emplace_back(std::move(arg.first), std::move(value));
+    }
+    return result;
+  }
+
+  tuple args_apply(const std::function<object(Box<object>&)>& func,
+      ArgsCont& cont) {
+    auto result = tuple(cont.size());
+    for(auto i = 0U; i < cont.size(); ++i) {
+      result[i] = func(cont[i]);
+    }
+    return result;
+  }
+
+  dict args_apply(const std::function<object(Box<object>&)>& func,
+      KwargsCont& cont) {
+    auto result = dict();
+    for(auto& arg : cont) {
+      result[arg.first] = func(arg.second);
+    }
+    return result;
+  }
+
+  class PythonLift {
+    public:
+      using Type = object;
+
+      PythonLift(object callable, args args, kwargs kwargs)
+        : m_callable(std::move(callable)),
+          m_args(args_apply([](object&& generator) {
```

Comment:
In this case you should be able to reduce this expression by eliminating the lambda entirely:

```m_args(apply(python_autobox<object>, std::move(args)))```

Reply:
Yep, it just used to be multi-line before, will change.

Reply:
Actually, here python_autobox<ReturnType, ArgFwd> uses perfect forwarding by its second template parameter. I can use it in this way: 
args_apply(python_autobox<object, object&&>, std::move(args)),
but I am not sure, is it acceptable or I should still prefer the lambda?

Reply:
Or python_autobox<object, object> for that matter.